libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/Plugins/SFNetworkPlugin/SFNetworkPlugin.m

```diff
@@ -0,0 +1,120 @@
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "SFNetworkPlugin.h"
+#import "CDVPlugin+SFAdditions.h"
+#import <SalesforceSDKCore/NSDictionary+SFAdditions.h>
+#import <SalesforceSDKCore/SFRestAPI+Blocks.h>
+
+// NOTE: must match value in Cordova's config.xml file.
+NSString * const kSFNetworkPluginIdentifier = @"com.salesforce.network";
+
+// Private constants.
+NSString * const kMethodArg       = @"method";
+NSString * const kPathArg         = @"path";
+NSString * const kEndPointArg     = @"endPoint";
+NSString * const kQueryParams     = @"queryParams";
+NSString * const kHeaderParams    = @"headerParams";
+NSString * const kfileParams      = @"fileParams";
+NSString * const kFileMimeType    = @"fileMimeType";
+NSString * const kFileUrl         = @"fileUrl";
+NSString * const kFileName        = @"fileName";
+NSString * const kPatchMethod     = @"PATCH";
+
+@implementation SFNetworkPlugin
+
+- (void) pgSendRequest:(CDVInvokedUrlCommand *) command
+{
+    NSDictionary *argsDict = [self getArgument:command.arguments atIndex:0];
+    SFRestMethod method = [SFRestRequest sfRestMethodFromHTTPMethod:[argsDict nonNullObjectForKey:kMethodArg]];
+    NSString* endPoint = [argsDict nonNullObjectForKey:kEndPointArg];
+    NSString* path = [argsDict nonNullObjectForKey:kPathArg];
+
+    /*
+     * Android's network library has a limitation that it does not support
+     * PATCH requests. However, Salesforce REST API does not allow POST for
+     * updates, it only allows PATCH. Hence, we work around the issue on
+     * Android by passing in method name PATCH as a URL parameter. However,
+     * we don't need this on iOS, since we can directly make a PATCH request.
+     */
+    if ([path containsString:kPatchMethod]) {
+        method = SFRestMethodPATCH;
+    }
+    NSDictionary* queryParams = [[NSDictionary alloc] init];
+    id queryParamsObj = [argsDict nonNullObjectForKey:kQueryParams];
+
+    /*
+     * Query params are NSDictionary for GET and encoded JSON in NSString
+     * for POST. These checks ensure that both cases are handled properly.
+     */
+    if ([queryParamsObj isKindOfClass:[NSString class]]) {
+        NSData* queryParamsData = [queryParamsObj dataUsingEncoding:NSUTF8StringEncoding];
+        queryParams = [NSJSONSerialization JSONObjectWithData:queryParamsData options:0 error:nil];
+    } else {
+        queryParams = queryParamsObj;
+    }
+    NSDictionary* headerParams = [argsDict nonNullObjectForKey:kHeaderParams];
+    NSDictionary* fileParams = [argsDict nonNullObjectForKey:kfileParams];
```

Comment:
Can we make use of the generics declarations here?  I think it would help document the structure coming in.